tools/tool-migration-utility/src/main/java/org/sagebionetworks/tool/migration/v4/MigrationClient.java

```diff
@@ -0,0 +1,496 @@
+		}
+	}
+	
+	/**
+	 * Migrate one type.
+	 * @param type
+	 * @param progress
+	 * @throws Exception 
+	 */
+	public DeltaData calculateDeltaForType(TypeToMigrateMetadata tm, String salt, long batchSize) throws Exception{
+
+		// First, we find the delta ranges
+		DeltaFinder finder = new DeltaFinder(tm, factory.createNewSourceClient(), factory.createNewDestinationClient(), salt, batchSize);
+		DeltaRanges ranges = finder.findDeltaRanges();
+		
+		// the first thing we need to do is calculate the what needs to be created, updated, or deleted.
+		// We need three temp file to keep track of the deltas
+		File createTemp = File.createTempFile("create", ".tmp");
+		File updateTemp = File.createTempFile("update", ".tmp");
+		File deleteTemp = File.createTempFile("delete", ".tmp");
+		
+		// Calculate the deltas
+		DeltaCounts counts = calculateDeltas(tm, ranges, batchSize, createTemp, updateTemp, deleteTemp);
+		return new DeltaData(tm.getType(), createTemp, updateTemp, deleteTemp, counts);
+		
+	}
+
+	/**
+	 * Calcaulte the deltas
+	 * @param type
+	 * @param batchSize
+	 * @param createTemp
+	 * @param updateTemp
+	 * @param deleteTemp
+	 * @throws SynapseException
+	 * @throws FileNotFoundException
+	 * @throws IOException
+	 */
+	private DeltaCounts calculateDeltas(TypeToMigrateMetadata typeMeta, DeltaRanges ranges, long batchSize, File createTemp, File updateTemp, File deleteTemp)	throws Exception {
+
+		BasicProgress sourceProgress = new BasicProgress();
+		BasicProgress destProgress = new BasicProgress();
+		BufferedRowMetadataWriter createOut = null;
+		BufferedRowMetadataWriter updateOut = null;
+		BufferedRowMetadataWriter deleteOut = null;
+		try{
+			createOut = new BufferedRowMetadataWriter(new FileWriter(createTemp));
+			updateOut = new BufferedRowMetadataWriter(new FileWriter(updateTemp));
+			deleteOut = new BufferedRowMetadataWriter(new FileWriter(deleteTemp));
+			
+			// Unconditional inserts
+			long insCount = 0;
+			List<IdRange> insRanges = ranges.getInsRanges();
+			for (IdRange r: insRanges) {
+				RangeMetadataIterator it = new RangeMetadataIterator(typeMeta.getType(), factory.createNewSourceClient(), batchSize, r.getMinId(), r.getMaxId(), sourceProgress);
+				RowMetadata sourceRow = null;
+				do {
+					sourceRow = null;
+					if (it.hasNext()) {
+						sourceRow = it.next();
+						createOut.write(sourceRow);
+						insCount++;
+					}
+				} while (sourceRow != null);
+			}
+			// Unconditional deletes
+			long delCount = 0;
+			List<IdRange> delRanges = ranges.getDelRanges();
+			for (IdRange r: delRanges) {
+				RangeMetadataIterator it = new RangeMetadataIterator(typeMeta.getType(), factory.createNewDestinationClient(), batchSize, r.getMinId(), r.getMaxId(), sourceProgress);
+				RowMetadata destRow = null;
+				do {
+					destRow = null;
+					if (it.hasNext()) {
+						destRow = it.next();
+						deleteOut.write(destRow);
+						delCount++;
+					}
+				} while (destRow != null);
+			}
+			

```

Comment:
Create the deltaCounts here, remove the threading and increment the counts on each loop.
Currently, it's only getting the last count.